Posture score chart UI

JordanSh opened the pull request:
This pr solves #2831 which is a part of #2755
the information text was moved down to be under the percentage number to save horizontal space.

![image](https://user-images.githubusercontent.com/51442161/151151106-085eb50b-563e-44c2-9191-7f43ac3c6756.png)

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_charts/cloud_posture_score_chart.tsx:
```diff
@@ -14,85 +14,114 @@ import {
 import { allNavigationItems } from '../../../common/navigation/constants';
 
+interface ChartName {
+  name?: string;
```
not all charts have name?

JordanSh replied:
umm good question, it's optional because when clicking on a benchmark score chart we want to use the name to filter the findings table. but in the general score, we don't want to filter the findings table with a name and rather show all findings.

i guess it can be done by passing query parameters? or maybe lifting up the `onElementClick` so you can pass the function (in this case the findings navigation function) from outside?

how would you go about it?

orouz replied:
i think lifting `onElementClick` sounds right

we can call it from the chart wrapper with the name (or more) and each consumer would decide if they use it or not. the main chart won't handler won't.  

in anyway, i think the main chart does have a name? if so, then the type shouldn't be optional

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_charts/cloud_posture_score_chart.tsx:
```diff
@@ -14,85 +14,114 @@ import {
+const getQuery = (name: ChartName['name'], evaluation: string): Query => ({
   language: 'kuery',
-  query: `rule.benchmark : "${benchmarkName}" and result.evaluation : "${evaluation}" `,
+  query: `${name ? `rule.benchmark : "${name}" and` : ''} result.evaluation : "${evaluation}"`,
```
i would love to build these with some query builder from `@kbn/es-query` but never managed to do that. 

we had an issue with this formatting. is it fixed here?

JordanSh replied:
you are right, i dont like this string either. i will try playing with it on another pr that I will open soon after this one is merged

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_charts/cloud_posture_score_chart.tsx:
```diff
@@ -14,85 +14,114 @@ import {
-      search: encodeQuery(
-        getBenchmarkAndResultEvaluationQuery(benchmarkName, rollupValue.toLowerCase())
-      ),
+      search: encodeQuery(getQuery(name, rollupValue.toLowerCase())),
```
consider renaming `getQuery` to indicate what query are we generating

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_charts/cloud_posture_score_chart.tsx:
```diff
@@ -14,85 +14,114 @@ import {
+    <Chart size={{ height: 75, width: 90 }}>
+      <Settings onElementClick={handleElementClick} />
+      <Partition
+        id={name || 'score_chart'}
```
if we're assuming it happens once (i think, as this is an `id`) then i'd make it explicit by making `score_chart` the `name` of the case where it is used.

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_charts/cloud_posture_score_chart.tsx:
```diff
@@ -14,85 +14,114 @@ import {
+        id={name || 'score_chart'}
+        data={data}
+        valueGetter="percent"
+        valueAccessor={(d: Datum) => d.value as number}
```
do we need to specify `Datum` here?

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_charts/cloud_posture_score_chart.tsx:
```diff
@@ -14,85 +14,114 @@ import {
+  totalPassed,
+  totalFindings,
+}: CloudPostureScoreChartProps['data']) => {
+  const percentage = `${Math.round(postureScore)}%`;
```
IMO it's best if FE doesn't change the score (even just rounding), but only display it. 
IIRC - BE does `toFixed`, so might as well do it all there in one go.

orouz replied:
if we keep it, i think it's better if this component just adds `%` and if the consumer would want to round it they'd pass it rounded

JordanSh replied:
can't agree here, BE gives a certain value and FE decides how to show it and what to do with it. we went from `toFixed(1)` to `round`, we might want to change it back, we might want to do calculations using the exact value.. its hard to tell, but it's good to have all the options.

orouz replied:
my thinking is that a different consumer of our api will have a different score than our UI, which might not be what we want

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_charts/cloud_posture_score_chart.tsx:
```diff
@@ -14,85 +14,114 @@ import {
+
+  return (
+    <EuiFlexGroup direction="column" justifyContent="flexEnd">
+      <EuiText style={{ fontSize: 36, fontWeight: 'bold', lineHeight: 1 }}>{percentage}</EuiText>
```
`EuiText` has a `size` prop, if possible it's probably best to use that instead of custom css

JordanSh replied:
size only goes up to `m` and above it you need to wrap `EuiText` and set another div to have a bigger font size and set the size to `inherit`.. I've kept the current implementation for the big text, using the size property for the smaller one.

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_charts/cloud_posture_score_chart.tsx:
```diff
@@ -14,85 +14,114 @@ import {
+const ComplianceTrendChart = () => <div>Trend Placeholder</div>;
+
+export const CloudPostureScoreChart = ({ data }: CloudPostureScoreChartProps) => {
+  if (data.totalPassed === undefined || data.totalFailed === undefined) return null;
```
i think the interface for these props is not optional?

JordanSh replied:
true, it was in the past. removed

orouz reviewed (dismissed):
see comments

JordanSh commented:
worth mentioning, I've changed `ChartPanel` a bit. it was supposed to handle viewing states for the approximately 6-7 charts we had in the old designs. now actually only have like 1-2. I've removed the empty state so it will be easier to pass components to it. when I'll get to #2757 the plan would be to remove `loading` and `error` states as well. since the entire dashboard will have to be one piece anyway. there is no point in loading anything individually and those viewing states will be managed from a higher component. at that point all `ChartPanel` will do is wrap charts with a `EuiPanel` and give them a header and some proper spacing.

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/dashboard_sections/summary_section.tsx:
```diff
@@ -7,42 +7,72 @@
+import { allNavigationItems } from '../../../common/navigation/constants';
+import { encodeQuery } from '../../../common/navigation/query_utils';
+
+const getQuery = (evaluation: string): Query => ({
```
i think `getEvaluationQuery` is more fitting
and i think `evaluation` is a string literal? `passed` | `failed` ?

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/dashboard_sections/summary_section.tsx:
```diff
@@ -7,42 +7,72 @@
+
+const getQuery = (evaluation: string): Query => ({
+  language: 'kuery',
+  query: `"result.evaluation : "${evaluation}"`,
```
does this have 3 `"` ? 🤔

JordanSh replied:
🙈

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/dashboard_sections/summary_section.tsx:
```diff
@@ -7,42 +7,72 @@
+
+    history.push({
+      pathname: allNavigationItems.findings.path,
+      search: encodeQuery(getQuery(rollupValue.toLowerCase())),
```
`rollupValue` is casted to `string`

consider checking it is before calling `toLowerCase()` to be on the safe side

JordanSh replied:
tolowercase actually not needed anymore, removed

orouz commented on x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/dashboard_sections/summary_section.tsx:
```diff
@@ -7,42 +7,72 @@
+          isError={getStats.isError}
+        >
+          <CloudPostureScoreChart
+            id="cloud_posture"
```
id need to be more specific ?

orouz reviewed (dismissed):
see comments
